Intel VT-D fixes

tbursztyka opened the pull request:
On going work to get intel vt-d not breaking things up.

Commit:
drivers/interrupt_controller: Enable xapic pass-through in intel vt-d

In XAPIC mode, it's possible to tell VT-D to let interrupt in compatible
format passing through.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

nashif commented:
> Shuch things is however not possible

Such things are however not possible....

Commit:
drivers/interrupt_controller: Make IOAPIC VT-D aware when relevant

If VT-D's interrupt remapping is in place, all IOAPIC RTEs need to get
remapped as well (or then they will be simply blocked).

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Make VT-D remap generic and handle flags

This will not only be used by MSI remapping but by all relevant
interrupts.

Fix also IRTE settings:
- handle x2apic for destination id
- destination mode is always logical (as for IOAPIC)

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Expanding VT-D public API

Adding a way to pre-install the irg/vector on an irte and a way to get
an irte based on irq/vector.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
- First commit seems to avoid the need for remapping ioapic (or MSI ones). I don't think we care about interrupt injections: zephyr is the sole user of the intel vt-d and does not manage VMs itself. (unless I miss something about the interrupt injection issue). But that works only on XAPIC mode, not on X2APIC so it's not really a fix.
- Trying to get ioapic properly remapped on further commits... still need more testing.

jukkar commented on drivers/interrupt_controller/intc_loapic.c:
```diff
@@ -69,6 +69,25 @@ static uint32_t loapic_device_power_state = DEVICE_PM_ACTIVE_STATE;
+	if (vtd != NULL) {
+		return true;
+	}
+#define DT_DRV_COMPAT intel_vt_d
```
Don't we need to take a backup `DRV_COMPAT_BAK` here, like what was done on `intc_ioapic.c`

tbursztyka replied:
Since there is any DRV_COMPAT being set for loapic it's unnecessary

jukkar replied:
Ok, just confirming.

Commit:
arch/x86: On irq remapping, all PCIE MSI/MSI-X need to be remapped

There is no need to differentiate with multi-vector or not, MSI vs
MSI-x: all need to be remapped if Intel VT-D is on.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
boards/x86: Support VT-D on UP squared relevantly

VT-D will only be enabled if MSI multi-vector will also be requested.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Fix VT-D EIME value

It's the 11th bit and not 11.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Fix VT-D remap destination on XAPIC

CPU-ID is written in a different format in case of XAPIC.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Fixing MSI vector allocation

Fixing an out of bound issue.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: All dynamic IRQ connection need to be remapped

Allocate an IRTE for all irq being connected through
arch_irq_connect_dynamic(). This will be mandatory since VT-D expects to
filter all interruptions (but the one it generates, as we will see
later).

Taking into account CONFIG_INTEL_VTD_ICTL_XAPIC_PASSTHROUGH, which could
help for debugging.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Make sure PCIE allocated IRTEs are tighten to irq/vector

As all interruption need to go through VT-D, calling vt-d remap will
happen on lower level as seen next, so make sure all pcie related
irq/vector get tighten to their respective allocated IRTE.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/pcie: Even single MSI based interrupt needs to be remapped

Refactor to handle this case. This is valid only when MSI multi-vector
feature is enabled.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
sys/arch: Add new interfaces for allocating and checking IRQs

Current usage was being limited to PCIE but there are new use case
that demand for this feature to be generalized.

It is made arch-specific since allocation policy may vary
per-architecture.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/pcie: Use the generic arch-interface for allocating IRQs

PCIE now uses the new interface.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Implement the IRQ allocation and usage interfaces for intel 64

This is the only architecture user for this at the moment.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: PCIE MSI vector allocator can use arch IRQ allocator

Instead of messing up with the PCI bus.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: PCIE MSI address and data may be out of remapping

In fact, in case of VT-D being enabled, it will require to get an
address and data for its own MSI based interrupts which cannot be
remapped (i.e.: will directly go to the relevant APIC).

This is necessary to get the Fault event supported in VT-D.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Add fault event support on Intel VT-D

IR faults are non-recoverable, so it's good to know why. Thus let's
handle the fault event and print the fault.

Other faults are printed as well.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Intel VT-D init should happen after ioapic

Or then it will badly crash trying to set up the fault event isr
handler.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Write IRTE as 64bits blocks in VT-D

Looks like a source of fault if pieces of IRTE are written instead of
the whole 64 bits block it belongs to.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Align VT-D helper function name

All are 32/64 differentiated, so does vtd_read_reg.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Adding QI support in VT-D

Looks like it's mandatory to invalidate the Interrupt Entry Cache in
VT-D and the only way to do so is to enable Queued Interface.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: VT-D IRTE table should be 4kb aligned

As specified by the IRTA location.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: VT-D may manipulate the cache

Depending on a VT-D capability, it might be necessary to flush objects
from the cache.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Simplify IRTE structure for readability

Let's avoid l/h accessors, and directly have all bits into one
structure.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Add a timeout check on VT-D QI operation

QI operation are not supposed to last forever.
It's not an actual timeout based on clock, but a dummy counter instead.
That's because system clock might not have beed initialized yet, since
VT-D's init comes first (and that same init will use QI...)

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Fix MMIO mapping

Those MMIO registers might be mapped in RAM and thus it's
really important to keep storage space for such mapping.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Use proper destination mode on VT-D

Let's always use logical Destination ID.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: No VT-D if no PCIE MSI multi-vector

There is no point to enable VT-D if PCIE MSI multi-vector support is not
requested, as VT-D is uniquely being used to remap MSI multi-vector
based interrupts.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/x86: Add VT-D node on apollo_lake

Kconfig options will anyway enable/disable the driver relevantly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/x86: Add VT-D node on Elkhart Lake

Kconfig options will anyway enable/disable the driver relevantly.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>
Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
boards/x86: Request VT-D support on ehl_crb relevantly

VT-D will only be enabled if MSI multi-vector will also be requested.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@jukkar @jhedberg @dcpleung please re-review. Finally got it working (getting the Intel DCI hw dbg was really helpful)

andyross commented on include/sys/arch_interface.h:
```diff
@@ -328,6 +328,29 @@ int arch_irq_connect_dynamic(unsigned int irq, unsigned int priority,
+ *
+ * @param irq the IRQ to declare being used
+ */
+void arch_irq_on_use(unsigned int irq);
```
"arch_irq_set_used(bool)" maybe?

tbursztyka replied:
you mean arch_irq_set_used(irq), yes its' a better name.

andyross commented on include/sys/arch_interface.h:
```diff
@@ -328,6 +328,29 @@ int arch_irq_connect_dynamic(unsigned int irq, unsigned int priority,
+ *
+ * @return The newly allocated IRQ or UINT_MAX on error.
+ */
+unsigned int arch_irq_allocate(void);
```
What's the difference between an IRQ that is "allocated" (which I assume means "has a x86 vector that won't be used by another driver"), "enabled" (can actually fire), "connected" (has a Zephyr handler registered) and "in use" (... honestly not sure?).

Right now we just have "connected" and "enabled".  If we're going to add more states to the model, we should be explicit about what they mean.  Isn't this something we could wrap together underneath irq_connect_dynamic()?

tbursztyka replied:
This is really a VT-D requirements: being able to allocate/connect outside of arch_irq_connect_dynamic() 

arch_irq_connect_dynamic() is the default irq connector (and it allocates vector underneath): in case of VT-D being enabled, the irq/vector will be remapped (it will allocate an VT-D IRTE and won't call z_x86_irq_connect_on_vector. Once irq enablement lands on IOAPIC side: it will recore the allocated irte and program the IOAPIC accordingly). And that's something we need to avoid for VT-D owns interrupt. (it cannot/shouldn't remap it's own interrupt). 

I was thinking that perhaps adding a new parameter or at least a dedicated flag could have been a solution: but arch_irq_connect_dynamic() is meant to be generic and this VT-D stuff is really an x86 specific technology so it's a no-go.

andyross reviewed (commented):
Some notes on the interrupt API.  I'm not nearly expert enough to critique the VT-D code but it looks very reasonable.

jhedberg commented on arch/x86/core/intel64/irq.c:
```diff
@@ -29,6 +29,28 @@ unsigned char _irq_to_interrupt_vector[CONFIG_MAX_IRQ_LINES];
+	if (vtd != NULL) {
+		return true;
+	}
+#define DT_DRV_COMPAT intel_vt_d
+	vtd = device_get_binding(DT_INST_LABEL(0));
+#undef DT_DRV_COMPAT
```
Seems like most/all VT-D code should be moved to a dedicated c-file so you don't need to do this kind of hacks with `DT_DRV_COMPAT`. At least my impression is that `DT_DRV_COMPAT` is supposed to be declared in the beginning of c-files and applicable to the entire file.

tbursztyka replied:
Actually, is there a way to avoid DT_DRV_COMPAT here? I get the usefulness of it in drivers/, but here when you need to get another device, what DTS macro should be used then?

Commit:
drivers/interrupt_controller: Fix command write in VT-D

A surprising issue: while sys_set_bit() on VTD_GCMD_REG work on
apollo_lake, it seems to create border effect on elkhart_lake: it may
reset some previous bit to 0.

So switching to full write of the register at once, which works for
both.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Add VT-D context cache invalidation

Not sure this is requered as we do not mangle with domains, but let's
make sure this cache is clear.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/pcie: Add a function to know MSI/MSI-x support of an endpoint

And since it does yet another round of pcie_get_cap() on PCIE_MSI_CAP_ID
and PCIE_MSIX_CAP_ID, let's factorize that into a utility function and
change the relevant places to use that function instead.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/pcie: Add a function for dynamic PCIe IRQ connection

The is meant to fix a chicken & egg issue with MSI interrupt remapping.
Currently, drivers first connect the irq (by-passing any possible MSI
remapping), so the IRQ ends-up being remapped at the IOAPIC level which
is not what we want.

So adding a dedicated function to properly handle this case. This is
valid only for runtime dynamic IRQ connection obviously.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/pcie: Add a marco for fixed or MSI IRQ

This is meant to be called by drivers on PCIe BUS with a fixed (from
DTS) IRQ or an MSI one. I.e.: all drivers without a runtime IRQ
detection/allocation.

The arch-specific macro is required to be implemented relevantly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Implementing ARCH_PCIE_IRQ_CONNECT for ia32 and intel64

While this does not change anything for ia32, this is fixing the chicken
& egg issue on intel64 if (and only if) interrupt remapping is enabled
so pcie_connect_dynamic_irq() is to be called properly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: Fixing the 2 drivers for PCIe IRQ connection

Now these 2 drivers should be working properly in case of IRQ remapping.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Fix IRTE delivery mode on VT-D

Only allow FIXED or LOW_PRIO modes.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Do not call irq controller on dedicated irq/vector function

This is meant for VT-D based MSI/MSI-x interrupt as well as VT-D's
internal interrupt: it does not require to get through the irq
controller.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Update PCIe MSI connection function

No flags is necessary for z_x86_irq_connect_on_vector().
Also, enable the irq through the controller if msi-x and no vt-d.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/pcie: Extending parameters to pcie_msi_map

n_vector will be necessary for VT-D actually.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Generate proper MSI address on VT-D

SHV bit depends on the number of vectors allocated.
If it's facing a multi-vector MSI array, it will set the bit.
If not the bit must be 0.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Add a function to retrieve ID from ACPI's DMAR

This will be necessary to get IOAPIC and HPET source ids for VT-D.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Add source id to VT-D interrupt remap

Change the API and apply that change where relevant.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/x86: Improve ACPI output on info test

Print out DMAR's device id for each device scopes.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/interrupt_controller: Add a VT-D option to disable src id check

Theoretically, source id check is not mandatory in the context we use
VT-D here (i.e. non VM one, just multi-vector MSI support)

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
hum, due to s/master/main or own side branch deletion (on github side that is), it seems like I am not able to re-open that PR. So I'll create another.
